catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+
+    private final Cache<String, CachedSubscription> subscriptionsCache;
+
+    private final Map<String, SubscriptionFactory> factories;
+
+    public SubscriptionContainerImpl(SubscriptionCacheLoader cacheLoader,
+            SubscriptionCacheWriter cacheWriter) {
+        this.subscriptionsCache = createCache(cacheLoader, cacheWriter);
+        this.factories = createFactoryMap();
+    }
+
+    /**
+     * On startup, load existing subscriptions from the central store.
+     * <p>
+     * This is an implementation detail. We need to ensure CRUD operations, loading the cache when
+     * Karaf starts, and the binding / unbinding of factories ... are ALL mutually exclusive.
+     * <p>
+     * Event handlers to remote cache events also fit into this category. We may need to consider
+     * passing around a shared lock if this logic needs to be split out. In addition, event handlers
+     * may need to invoke this class asynchronously if they are not asynchronous themselves, otherwise
+     * we can incur deadlock by contesting the cache implementation's locking mechanism.
+     * <p>
+     * If event handlers are given this container impl, redundant writes and deletes to the backing
+     * store can <b>NOT</b> be considered errors.
+     *
+     * @see SubscriptionCacheLoader#loadAll(Iterable)
+     */
+    public synchronized void init() {
+        subscriptionsCache.loadAll(Collections.emptySet(),
+                false,
+                new CacheUpdateCompletionListener());
+
+        StreamSupport.stream(subscriptionsCache.spliterator(), false)
+                .map(Cache.Entry::getValue)
+                .filter(CachedSubscription::isNotRegistered)
+                .filter(sub -> factories.containsKey(sub.getMetadata()
+                        .getTypeName()))
+                .forEach(sub -> sub.registerSubscription(factories.get(sub.getMetadata()
+                        .getTypeName())));
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * An attempt is made to return the subscription object so long as:
+     * <ol>
+     * <li>The subscription being identified exists in the cache</li>
+     * <li>The subscription's type matches the identifier's type</li>
+     * <li>The subscription is registered with OSGi</li>
+     * </ol>
+     * <p>
+     * One impact of an incorrect implementation of {@link SubscriptionFactory} will be incorrect
+     * {@code null} results returned for client traffic that occurs during service initialization time.
+     * This is because queries for non-registered subscriptions that do exist but haven't been initialized
+     * yet return null. Subscription container clients that expose external services must ensure their
+     * factory is available prior to, or at the same time as, those exposed services.
+     *
+     * @see SubscriptionFactory
+     */
+    @Nullable
+    @Override
+    public synchronized Subscription get(SubscriptionIdentifier identifier) {
```

Comment:
❗️ General comments on the `synchronized` keyword usage. Instead of synchronizing the whole methods, we should consider using `synchronized(this)` and only synchronize the blocks of code that are not thread-safe.